### DISCUSSION TITLE
Renovate doesn't detect dependencies in gradle custom configurations

### DISCUSSION THREAD

**Discussion by eirnym** [renovatebot/renovate#20122] [Suggest an Idea]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

I use Spotless as part of my code quality CI, and keep a special configuration for it. I have a special configuration for few reasons:
* Keep used formatters for spotless up-to-date.
* Allow tools such as Renovate to update these dependencies
* These are not project's dependencies, so I'd prefer to keep these dependencies aside.

These definitions are quite common and used for various reasons, such as dependencies for a tool used in a plugin. So I wish to allow Renovate rules to pick up these dependency definitions. 

My goal **is not** to separate these configurations in the renovate's configuration, it will be too much to work with.  I prefer to have at least basic support for in the near future and _maybe_ to have some improvement later on.

Most common configurations used in projects are `implementation` and `api` (for Java/Kotlin/Groovy applications these configurations are created by `java` or `java-library` plugins). Other configurations  can be created on demand by a `build.gradle*` script or a plugin. Any dependency declaration in Gradle has two most common formats as follows:

* `configuration(dependencySpec)` 
* `add("configurationName", dependencySpec)`

Below is a custom configuration example, which I assume, that Renovate will pick up.

```kotlin
val styleCheckers: Configuration by configurations.creating

dependencies {
    styleCheckers("com.pinterest:ktlint:$ktLintFormatVersion")
    styleCheckers("com.palantir.javaformat:palantir-java-format:$palantirJavaFormatVersion")
}
```

### Relevant debug logs

<details><summary>Logs</summary>

```
I haven't found any relevant logs. Probably it's just not implemented or implementation hardcoded. Search terms are:
styleCheckers, ktlint, palantir

https://gist.github.com/eirnym/71b0ee4c01db93b89ec4d6cc5366e0df
```

</details>

### Have you created a minimal reproduction repository?

No reproduction, but I have linked to a public repo where it occurs

**Comment by eirnym** [comment]
repo: https://github.com/jsonschema2dataclass/js2d-gradle

**Comment by rarkins** [comment]
Converted to an Ideas discussion. For comment by @Churro @zharinov

**Comment by Churro** [comment]
Renovate works nicely with any configuration you specify, even the custom `stylecheckers`. 

The issue in the reproduction repo you linked is that variables with version values are provided by a composite build (feature request: https://github.com/renovatebot/renovate/issues/9410). [build.gradle.kts](https://github.com/jsonschema2dataclass/js2d-gradle/blob/main/build.gradle.kts) looks like this:

```kt
import org.jsonschema2dataclass.internal.plugin.palantirJavaFormatVersion
...
styleCheckers("com.palantir.javaformat:palantir-java-format:$palantirJavaFormatVersion")
```

and would need to map to [SpotlessPlugin.kt](https://github.com/jsonschema2dataclass/js2d-gradle/blob/main/internal/src/main/kotlin/org/jsonschema2dataclass/internal/plugin/SpotlessPlugin.kt#L10):
```kt
const val palantirJavaFormatVersion = "2.21.0"
```

Your specific problem may or may not have the same root cause. In case your code looks the same as the linked one, just hard-code the version directly or in a local variable and renovate will take care of upgrading the dependency.

**Comment by eirnym** [comment]
I see. I can't move these version numbers to the main `build.gradle.kts` for few reasons:
* I can't just apply the plugin (it's java 11 only, few of my builds require Java 8)
* Plugin doesn't use configurations.
* I don't want to have a mess of if-then-elses to keep plugin configuration in the same main `build.gradle.kts`.
* I prefer to have a single place to pin the version.
* These dependencies is not my code

Is there any way in Renovate to fulfill these requirements at the moment, before #9410 will be fixed?

**Reply by Churro** [reply]
As long as the version is defined somewhere in your project, you can [craft a pattern](https://docs.renovatebot.com/modules/manager/regex/) for `regexManagers` and match it that way.

**Reply by eirnym** [reply]
Thank you, it's an interesting proposition, I'll definitely check it out! 

The question left: How do I check if configuration is correct before pushing it to `main` branch or installing Renovate locally?

**Reply by Churro** [reply]
Depends on what you expect from the check. To just validate the JSON syntax, `jq type renovate.json` is enough. To see what renovate does with your config, you could run it like this: `renovate --dry-run`

**Reply by eirnym** [reply]
Thank you a lot! It finally worked
